### ISSUE TITLE
Json serialization of Maps

### ISSUE THREAD

**Issue by nooone** [libgdx/libgdx#3030]
The current implementation of `Json` has a special handling for map-like datastructures. It tries to serialize `java.util.Map`, `com.badlogic.gdx.utils.ArrayMap` and `com.badlogic.gdx.utils.ObjectMap` in special ways.

As always with the libgdx collection framework, there are more classes that are not covered in a generic way (like `ObjectIntMap` for example). I guess that's a problem that cannot be solved though.

However, it's good that at least some standard classes are covered. On the other hand, they are serialized in a pretty weird way. It looks like the map entries are serialized as fields with "key=value". If I saw it right, it's possible to configure the value type serialization by setting the `elementType`. It's not possible though to change how the keys are serialized. The keys are just converted to a string as the fields name ([code](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/Json.java#L516)).

That works sometimes (in case there is a simple String representation, like for Enum, Integer, String itself etc). However they are also read back as Strings, since there is no way for Json to know which type is expected (because of type erasure).

A simple example:

	ObjectMap<Integer, Integer> numbers = new ObjectMap<Integer, Integer>();
	numbers.put(10, 5);
	String jsonString = json.toJson(numbers);
	System.out.println(jsonString);
	numbers = json.fromJson(ObjectMap.class, jsonString);
	System.out.println(numbers);
	System.out.println(numbers.get(10));

Will result in this:

	{10:{class:java.lang.Integer,value:5}}
	{10=5}
	null

That last `null` is unexpected. That's because in our `ObjectMap<Integer, Integer>`, we now have a *`String`* key instead of `Integer`. Type erasure at work.

I'd suggest to change the current serialization to actually serialize the entries of the map as an array of entries instead of key=value fields.

[Here](http://pastebin.com/0h5yVNdU) is an implementation of a `ObjectMapSerializer` which makes use of that approach and is able to serialize and deserialize the maps correctly.

I think this is a much better way of doing it. If you agree, I'll try to make a PR for that.

**Comment by NathanSweet** [comment]
JSON keys are always strings. If we know the concrete type of the key *and* the type can be represented as a string we can do:

```
{ 123: blah }
```

If we don't know the concrete type of the key, we need a way to write the type in the JSON, something like:

```
[
   { class: Integer, key: 123, value: blah }
]
```

If the key was an object:

```
[
   { class: Stuff, key: { value:123, other:stuff }, value: blah }
]
```

If we don't know the type of the map, we need that in the JSON too, so if we use an array for the entries we need another container:

```
{
   class: TreeMap,
   entries: [
      { class: Stuff, key: { value:123, other:stuff }, value: blah }
   ]
}
```

This serializes the data, but doesn't really look like the data structure being serialized anymore.

Keep in mind you can always write your own serializer for any type you like.

**Comment by nooone** [comment]
@NathanSweet I'm not exactly sure what the "message" of your post is. All you say is correct, but I don't understand what you think the consequences should be.

What I wanted to say is that the current handling of the maps, predefined in `Json` is not much better than not having any special handling, because the only case in which it works is, when the keys of the map are of type `String`.

The serialization works also in other cases, for example for Enums or Integers, but when deserializing it, there's no way to tell it which type it should use for the keys, as `elementType` is used for the values and there is no such configuration for the keys.

> Keep in mind you can always write your own serializer for any type you like.

Yes, that's why I wrote my own serializer for this case. But if libgdx already has some special implementation, it should at least be useful, don't you think? :)

**Comment by NathanSweet** [comment]
Yeah I just rambled some stuff as I had only a few free minutes. My point was that it's not easy to describe keys for maps that can't be represented as string OR when the concrete type of the key is not known. For the case when the value **can** be represented by a JSON object key (which is always string) AND we know the concrete type, then I think the Json class should be able to handle the type conversion and the JSON output won't turn into the junk I posted above. If you want to make a PR to do that, that would be super! :)
